core/services/relay/evm/cap_encoder.go

```diff
@@ -0,0 +1,91 @@
+package evm
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	consensustypes "github.com/smartcontractkit/chainlink-common/pkg/capabilities/consensus/types"
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	abiutil "github.com/smartcontractkit/chainlink/v2/core/chains/evm/abi"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/types"
+)
+
+const (
+	abiConfigFieldName = "abi"
+	userEncoderName    = "user"
+	outerEncoderName   = "outer"
+)
+
+// TODO update in common
+const WorkflowIDFieldName = "INTERNAL_worklow_id"
+const WorkflowExecutionIDFieldName = "INTERNAL_workflow_execution_id"
+const InnerUserDataFieldName = "INTERNAL_user_data"
+
+type capEncoder struct {
+	codec commontypes.RemoteCodec
+}
+
+var _ consensustypes.Encoder = (*capEncoder)(nil)
+
+func NewEVMEncoder(config *values.Map) (consensustypes.Encoder, error) {
+	// parse the "inner" encoder config - user-defined fields
+	wrappedSelector, err := config.Underlying[abiConfigFieldName].Unwrap()
+	if err != nil {
+		return nil, err
+	}
+	selectorStr, ok := wrappedSelector.(string)
+	if !ok {
+		return nil, fmt.Errorf("expected %s to be a string", abiConfigFieldName)
+	}
+	selector, err := abiutil.ParseSignature("inner(" + selectorStr + ")")
+	if err != nil {
+		return nil, err
+	}
+	userEncoderConfig, err := json.Marshal(selector.Inputs)
+	if err != nil {
+		return nil, err
+	}
+
+	// prepare the "outer" encoder config - a tuple of (user_fields, workflow_id, workflow_execution_id)
+	outerSelector, err := abiutil.ParseSignature("outer(" +
+		InnerUserDataFieldName + " bytes," +
+		WorkflowIDFieldName + " string," +
+		WorkflowExecutionIDFieldName + " string)")
+	if err != nil {
+		return nil, err
+	}
+	outerEncoderConfig, err := json.Marshal(outerSelector.Inputs)
+	if err != nil {
+		return nil, err
+	}
+
+	codecConfig := types.CodecConfig{Configs: map[string]types.ChainCodecConfig{
+		userEncoderName:  {TypeABI: string(userEncoderConfig)},
+		outerEncoderName: {TypeABI: string(outerEncoderConfig)},
+	}}
+	c, err := NewCodec(codecConfig)
```

Comment:
@cedric-cordenier regarding our discussion about getting codec form a Provider - here I am creating a new object from CodecConfig and provider gives me a ready one. How do I plug in my config then?

Reply:
Each provider supplies one.  Currently, it's only in median (chain reader is to be tested there first).

It's set up here: https://github.com/smartcontractkit/chainlink/blob/develop/core/services/relay/evm/evm.go#L509  

and returned here:
https://github.com/smartcontractkit/chainlink/blob/develop/core/services/relay/evm/evm.go#L509

For some context, the median plugin then makes use of it (if it's enabled) here

https://github.com/smartcontractkit/chainlink-feeds/blob/master/median/plugin.go#L53

Reply:
@bolekk Hmmmm... I see what you mean now. We want to instantiate encoders dynamically, but only get a single provider per job since providers are instantiated as part of job creation.

In the immediate term I think we need to move this into the capability as we discussed yesterday. Long-term, we need to expose a Relayer API to the loopp, so that it can create providers on-demand.

Reply:
There is a loop ability for it [here](https://github.com/smartcontractkit/chainlink-common/blob/main/pkg/loop/internal/codec.go).  IIRC, the relay is actually running EVM and not LOOP, but you can run it as a LOOP with that config.

Reply:
we decided to provide a simple codec factory to the OCR capability for now (in another PR)